core/modules/wpublish/QueryChunkStatistics.h

```diff
@@ -0,0 +1,170 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+#ifndef LSST_QSERV_WSCHED_QUERYCHUNKSTATISTICS_H
+#define LSST_QSERV_WSCHED_QUERYCHUNKSTATISTICS_H
+
+// System headers
+
+// Qserv headers
+#include "wbase/Task.h"
+
+namespace lsst {
+namespace qserv {
+namespace wpublish {
+
+
+class QueryChunkStatistics;
+
+
+/// Statistics for a single user query.
+class QueryStatistics {
+public:
+    using Ptr = std::shared_ptr<QueryStatistics>;
+
+    explicit QueryStatistics(QueryId const& queryId) : _queryId{_queryId} {}
+
+    void addTask(wbase::Task::Ptr const& task);
+
+    bool isDead(std::chrono::seconds deadTime, std::chrono::system_clock::time_point now);
+
+    friend class QueryChunkStatistics;
+
+private:
+    bool _isMostlyDead();
+
+    std::mutex _mx;
+    QueryId const _queryId;
+    std::chrono::system_clock::time_point _touched{std::chrono::system_clock::now()};
+
+    int _size{0};
+    int _tasksCompleted{0};
+    int _tasksRunning{0};
+    int _tasksBooted{0}; ///< Number of Tasks booted for being too slow.
+
+    double _totalCompletionTime{0.0};
+
+    std::map<int, wbase::Task::Ptr> _taskMap;
+};
+
+/// Statistics for a table in a chunk. Statistics are base on the slowest table in a query,
+/// so this most likely includes values for queries on _scanTableName and queries that join
+/// against _scanTableName. Source is slower than Object, so joins with Source and Object will
+/// have there statistics logged with Source.
+class ChunkStatsTable {
+public:
+    using Ptr = std::shared_ptr<ChunkStatsTable>;
+
+    static std::string makeTableName(std::string const& db, std::string const& table) {
+        return db + ":" + table;
+    }
+
+    ChunkStatsTable(int chunkId, std::string const& name) : _chunkId{chunkId}, _scanTableName{name} {}
+
+    void addTaskFinished(double duration);
+
+private:
+    std::mutex _mtx;
+    int const _chunkId;
+    std::string const _scanTableName;
+
+    std::uint64_t _tasksCompleted{0}; ///< Number of Tasks that have completed on this chunk/table.
+    std::uint16_t _tasksBooted{0}; ///< Number of Tasks that have been booted for taking too long.
+
+    double _avgCompletionTime{0.0}; ///< weighted average of completion time
+    double _weightAvg{49.0}; ///< weight of previous average
+    double _weightDur{1.0}; ///< weight of new measurement
+    double _weightSum{_weightAvg + _weightDur}; ///< denominator
+};
+
+
+/// Statistics for one chunk.
+class ChunkStats {
+public:
+    using Ptr = std::shared_ptr<ChunkStats>;
+
+    ChunkStats(int chunkId) : _chunkId{chunkId} {}
+
+    ChunkStatsTable::Ptr add(std::string const& scanTableName, double duration);
```

Comment:
I'd make `duration` dimensional (like milliseconds) or at least change name to have _sec or _minutes suffix and comment it.